relays/finality-relay/src/finality_loop.rs

```diff
@@ -0,0 +1,563 @@
+			Some(Some(finality_proof)) => finality_proof,
+			_ => break,
+		};
+		let finality_proof_target_header_number = match finality_proof.target_header_number() {
+			Some(target_header_number) => target_header_number,
+			None => {
+				continue;
+			}
+		};
+
+		let justified_header =
+			prune_unjustified_headers::<P>(finality_proof_target_header_number, &mut unjustified_headers);
+		if let Some(justified_header) = justified_header {
+			recent_finality_proofs.clear();
+			selected_finality_proof = Some((justified_header, finality_proof));
+		} else {
+			// the number of proofs read during single wakeup is expected to be low, so we aren't pruning
+			// `recent_finality_proofs` collection too often
+			recent_finality_proofs.push_back((finality_proof_target_header_number, finality_proof));
+		}
+	}
+
+	// remove obsolete 'recent' finality proofs + keep its size under certain limit
+	let oldest_finality_proof_to_keep = selected_finality_proof
+		.as_ref()
+		.map(|(header, _)| header.number())
+		.unwrap_or(best_number_at_target);
+	prune_recent_finality_proofs::<P>(
+		oldest_finality_proof_to_keep,
+		recent_finality_proofs,
+		sync_params.recent_finality_proofs_limit,
+	);
+
+	Ok(selected_finality_proof)
+}
+
+pub(crate) fn prune_unjustified_headers<P: FinalitySyncPipeline>(
+	justified_header_number: P::Number,
+	unjustified_headers: &mut UnjustifiedHeaders<P>,
+) -> Option<P::Header> {
+	prune_ordered_deque(justified_header_number, unjustified_headers, usize::MAX, |header| {
+		header.number()
+	})
+}
+
+pub(crate) fn prune_recent_finality_proofs<P: FinalitySyncPipeline>(
+	justified_header_number: P::Number,
+	recent_finality_proofs: &mut FinalityProofs<P>,
+	recent_finality_proofs_limit: usize,
+) {
+	prune_ordered_deque(
+		justified_header_number,
+		recent_finality_proofs,
+		recent_finality_proofs_limit,
+		|(header_number, _)| *header_number,
+	);
+}
+
+fn prune_ordered_deque<T, Number: relay_utils::BlockNumberBase>(
+	header_number: Number,
+	ordered_deque: &mut VecDeque<T>,
+	maximal_deque_size: usize,
+	extract_header_number: impl Fn(&T) -> Number,
+) -> Option<T> {
+	// `VecDeque::binary_search_by_key` is not stabilized. Let's call `make_contiguous` and use
+	// `slice::binary_search_by_key` instead
+	let position = ordered_deque
+		.make_contiguous()
+		.binary_search_by_key(&header_number, extract_header_number);
+
+	// first extract element we're interested in
+	let extracted_element = match position {
+		Ok(position) => {
+			let updated_deque = ordered_deque.split_off(position + 1);
+			let extracted_element = ordered_deque.pop_back().expect(
+				"binary_search_by_key has returned Ok(); so there's element at `position`;\
+					we're splitting deque at `position+1`; so we have pruned at least 1 element;\
+					qed",
+			);
+			*ordered_deque = updated_deque;
+			Some(extracted_element)
+		}
+		Err(position) => {
+			*ordered_deque = ordered_deque.split_off(position);
+			None
+		}
+	};
```

Comment:
Since the header found here is only used by `prune_unjustified_headers()` I would move this code there. Then, since the splitting code is pretty short, I'd just inline that in `prune_recent_finality_proofs()` and `prune_unjustified_headers()`